### PR TITLE
Improve Ubuntu/Debian container configs, add Ubuntu 20.04 and update buildtools

### PR DESCRIPTION
**Pull request by pbrkr**
* Improve the Debian & Ubuntu container configs to resolve issues seen with my use cases.
* Add a Dockerfile for Ubuntu 20.04
* Updated to buildtools-extended tarball from Yocto 3.1 for Centos 7 & Ubuntu 16.04.

### CODE REVIEW

**Comment by pbrkr** [comment]
@rewitt1: Ping.

**Comment by rewitt1** [comment]
@pbrkr Sorry this has been sitting for so long. I was on leave. I'll review it today.

File: dockerfiles/ubuntu/ubuntu-16.04/ubuntu-16.04-base/Dockerfile [hunk 1]
```diff
@@ -17,7 +17,7 @@
 FROM ubuntu:16.04
 
 RUN apt-get update && \
-    apt-get install -y \
+    DEBIAN_FRONTEND=noninteractive apt-get install -y \
```
Comment (rewitt1) [inline dockerfiles/ubuntu/ubuntu-16.04/ubuntu-16.04-base/Dockerfile:20]: Is this needed due to the tzdata addition? If so, could you add the reason to the commit message?

Comment (pbrkr) [reply dockerfiles/ubuntu/ubuntu-16.04/ubuntu-16.04-base/Dockerfile:20]: The `tzdata` addition triggered a failure without this but in general this should be there incase other packages acquire interactive questions in the future.

File: dockerfiles/ubuntu/ubuntu-16.04/ubuntu-16.04-base/Dockerfile [hunk 2]
```diff
@@ -27,6 +27,7 @@ RUN apt-get update && \
         sysstat \
         texinfo \
         gcc-multilib \
+        g++-multilib \
```
[the remark this answers is not in the document]
Comment (pbrkr) [reply dockerfiles/ubuntu/ubuntu-16.04/ubuntu-16.04-base/Dockerfile:30]: It's going to be difficult to give exact failure cases without investing several hours making them appear again. This is needed if you've got some 32-bit host tool being built which uses C++ which you find when you're building with the Arago distro and a pre-built toolchain.

File: dockerfiles/ubuntu/ubuntu-16.04/ubuntu-16.04-base/Dockerfile [hunk 3]
```diff
@@ -48,13 +49,18 @@ RUN apt-get update && \
     echo "" | vncpasswd -f > /etc/vncskel/.vnc/passwd && \
     chmod 0600 /etc/vncskel/.vnc/passwd && \
     useradd -U -m yoctouser && \
-    /usr/sbin/locale-gen en_US.UTF-8
+    /usr/sbin/locale-gen en_US.UTF-8 && \
```
[the remark this answers is not in the document]
Comment (pbrkr) [reply dockerfiles/ubuntu/ubuntu-16.04/ubuntu-16.04-base/Dockerfile:52]: I can update the commit message that's no issue.

The LANG variable not being set trips up some host tools written in Perl IIRC. At this point it's more of a heuristic that "weird, hard to debug build failures can occur if LANG isn't set" than an actual recent case I've seen fail.

File: dockerfiles/ubuntu/ubuntu-20.04/ubuntu-20.04-base/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,67 @@
+# ubuntu-20.04-base
+# Copyright (C) 2015-2018 Intel Corporation
```
[the remark this answers is not in the document]
Comment (pbrkr) [reply dockerfiles/ubuntu/ubuntu-20.04/ubuntu-20.04-base/Dockerfile:2]: I copy-pasted from the previous Dockerfile without modifying the header.

File: tests/container/smoke.sh [hunk 5]
```diff
@@ -45,7 +45,7 @@ chmod -R 777 $workdir
 ${ENGINE_CMD} run -t --rm --entrypoint=/bin/bash -u $(id -u):$(id -g) -w '/' $image -c 'ls /usr/bin/dumb-init'
 
 # Try to build quilt-native which is small and fast to build.
-git clone --depth 1 --branch master \
+git clone --depth 1 --branch dunfell \
```
[the remark this answers is not in the document]
Comment (pbrkr) [reply tests/container/smoke.sh:48]: My thought here is that if a Travis build fails on a new pull request the author will think their change was bad. If we're following the master branch of poky the build failure may be due to upstream churn though.

[the remark this answers is not in the document]
Comment (rewitt1) [reply tests/container/smoke.sh:48]: Yeah I'm not really sure the best solution for this. Honestly if we could store the last known good version and fallback if HEAD fails, then that should address the problem. However, we'd also have to add some check to make sure that if the good version gets stale that it gets investigated.

I'd have to investigate how to do that though.

File: dockerfiles/ubuntu/ubuntu-16.04/ubuntu-16.04-base/Dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (rewitt1) [reply dockerfiles/ubuntu/ubuntu-16.04/ubuntu-16.04-base/Dockerfile:30]: Alright, then if you could just put "This is needed if you've got some 32-bit host tool being built which uses C++." in the body, that's sufficient.

File: dockerfiles/ubuntu/ubuntu-16.04/ubuntu-16.04-base/Dockerfile (hunk 3 above)
[the remark this answers is not in the document]
Comment (rewitt1) [reply dockerfiles/ubuntu/ubuntu-16.04/ubuntu-16.04-base/Dockerfile:52]: We should probably do then LANG/tz stuff it in the other distros as well then. I try to keep them as symmetric as possible with regards to configuration. But I'm ok with merging this for now as long as we create an issue for the other distros.

File: dockerfiles/ubuntu/ubuntu-20.04/ubuntu-20.04-base/Dockerfile (hunk 4 above)
[the remark this answers is not in the document]
Comment (rewitt1) [reply dockerfiles/ubuntu/ubuntu-20.04/ubuntu-20.04-base/Dockerfile:2]: I can't tell if you're going to change it, or if you're just telling me why it's that way. :)

[the remark this answers is not in the document]
Comment (pbrkr) [reply dockerfiles/ubuntu/ubuntu-20.04/ubuntu-20.04-base/Dockerfile:2]: I can change it :)

File: dockerfiles/ubuntu/ubuntu-16.04/ubuntu-16.04-base/Dockerfile (hunk 1 above)
Comment (rewitt1) [reply dockerfiles/ubuntu/ubuntu-16.04/ubuntu-16.04-base/Dockerfile:20]: Ok, then could put that in the body of the message?

File: tests/container/smoke.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (pbrkr) [reply tests/container/smoke.sh:48]: I'll drop this from the series for now and open an issue to track it.

File: dockerfiles/ubuntu/ubuntu-16.04/ubuntu-16.04-base/Dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (pbrkr) [reply dockerfiles/ubuntu/ubuntu-16.04/ubuntu-16.04-base/Dockerfile:30]: I've kicked off a build as I wasn't happy with how vague I was being. I'll see what it turns up and let you know.

File: dockerfiles/ubuntu/ubuntu-16.04/ubuntu-16.04-base/Dockerfile (hunk 1 above)
Comment (pbrkr) [reply dockerfiles/ubuntu/ubuntu-16.04/ubuntu-16.04-base/Dockerfile:20]: Will do.

File: dockerfiles/ubuntu/ubuntu-16.04/ubuntu-16.04-base/Dockerfile (hunk 3 above)
[the remark this answers is not in the document]
Comment (pbrkr) [reply dockerfiles/ubuntu/ubuntu-16.04/ubuntu-16.04-base/Dockerfile:52]: Ok, I'll open an issue before this is merged.

[the remark this answers is not in the document]
Comment (pbrkr) [reply dockerfiles/ubuntu/ubuntu-16.04/ubuntu-16.04-base/Dockerfile:51]: I've dropped this for now until I replicate the issues seen due to lack of tzdata. LANG is fine as it's set explicitly in runbitbake.py, poky-container and extsdk-container.

**Comment by pbrkr** [comment]
@rewitt1: The build is failing due to exactly the issue captured in #44. It looks like Yocto master branch no longer builds on Ubuntu 16.04 without using buildtools.

**Comment by pbrkr** [comment]
@rewitt1: All green now, sorry for the noise there.

File: dockerfiles/ubuntu/ubuntu-16.04/ubuntu-16.04-base/Dockerfile [hunk 6]
```diff
@@ -52,6 +52,13 @@ RUN apt-get update && \
     echo 'dash dash/sh boolean false' | debconf-set-selections && \
     DEBIAN_FRONTEND=noninteractive dpkg-reconfigure dash
 
+# The versions of gcc, tar and other tools found in Ubuntu 16.04 are now too
+# old to build recent Yocto Project releases. Install the buildtools tarball to
+# provide the required tools.
+COPY install-buildtools.sh /
+RUN bash /install-buildtools.sh && \
+    rm /install-buildtools.sh
+
```
[the remark this answers is not in the document]
Comment (pbrkr) [reply dockerfiles/ubuntu/ubuntu-16.04/ubuntu-16.04-base/Dockerfile:61]: That may have got mixed up, I rebased the patches a few times.

Sorry for the late reply, gmail must have dropped the notification email for your comment there.

**Comment by pbrkr** [comment]
@rewitt1: Sorry, looks like gmail dropped the notification email for your comments 12 days ago so I never replied.

poky-container will need updating now to include an Ubuntu 20.04 image and to enable buildtools if it's installed.

**Pull request merged**